### PR TITLE
[VT]: Add Virtual Terminal Object Classes

### PR DESCRIPTION
**Pull request by ad3154**
This is intended to be used with a VT server (coming soon? ) or with an object pool editor program. I wanted to open a PR for the ones I have done before the file gets even larger.

- Added classes to represent most VT objects.

- [x] Todo: Unit tests
- [x] Todo: Get Attribute Value functions

Supported objects so far:
- Working Set
- Data Mask
- Alarm Mask
- Container
- Soft Key mask
- Key
- Key Group
- Button
- Input Boolean
- Input String
- Input Number
- Input List
- Output String
- Output Number
- Output List
- Output Line
- Output Rectangle
- Output Ellipse
- Output Polygon
- Output Meter
- Output Linear Bar Graph
- Output Arched Bar Graph
- Picture Graphic
- Number Variable
- String Variable
- Font Attributes
- Line Attributes
- Fill Attributes
- Input Attributes
- Extended Input Attributes
- Object Pointer
- External Object Pointer
- Macro
- Colour Map
- Window Mask

### CODE REVIEW

**Comment by ad3154** [comment]
@GwnDaan I've finished my todos on this for now, so it's ready to review at your convenience.

Despite the amount of code, it's mostly just setters and getters for the various object types, so there's not really that much functionality to worry about. The object pool deserializer and whatnot will come in another PR.

**Comment by GwnDaan** [comment]
> @GwnDaan I've finished my todos on this for now, so it's ready to review at your convenience.
> 
> Despite the amount of code, it's mostly just setters and getters for the various object types, so there's not really that much functionality to worry about. The object pool deserializer and whatnot will come in another PR.

Sounds good! I'll try to review it in the coming days, and then I think I'll just mostly look at the tests and not so much at the rest, as you suggested👍

File: isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp
Comment (GwnDaan) [inline isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp:296]: I'm not sure how this can work? Shouldn't you want to simply overwrite the attribute of let's say an output string to achieve this? 

Consider the following (incomplete) pool: 
- Output string (id=20)
- String variable (id=30)
- String variable (id=31)

Then you'd simply want to change the variable reference of the "Output string (id=20)" from "30" to "31"? Instead of replacing things in the object pool like done here?

File: isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp
Comment (ad3154) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp:304]: That's fair - the way I'm consuming these classes myself is safe I think but that doesn't mean it would be for others. Might be better to store a weak pointer or something

Comment (GwnDaan) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp:304]: Or maybe just require an argument in the functions that use it? It doesn't seem to be used that often IIRC

File: isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp
Comment (ad3154) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp:296]: Well, it's not actually replacing an object, per-se, so maybe the name could be clearer... let me explain from the beginning.

So, some of the reason why this exists was to try and reduce the amount of functions and variables in each object. The base object has a `std::vector<ChildObjectData>`, so I've tried to make use of that as much as possible. So if an object like an input number can only ever have 1 of certain kinds of references (number variable, and font attributes in this case) I just put them in that vector instead of making special setters and getters for them. I don't know if this is a good thing or a bad thing, but since I have to do a lot of iterating over objects' children recursively in the VT server, it's maybe more consistent? 

Of course then I still had to support the case where someone wanted to change that child, especially something like a `FontAttributes`, so I made this function to do it for me. It replaces the object ID of the old object in the child vector with the object id of the new child in the vector. So it doesn't actually "replace" something in the pool, if you know what I mean, it's just changing the object ID of a child to another object's ID.

So, I guess I think this function is doing as you said: `change the variable reference of the "Output string (id=20)" from "30" to "31"` it's just a little less than obvious.

If you think it would be clearer to rename it, I could do that. Or if you would prefer concrete child variables with named setters and getters I could do that too, it'll just mean a longer file.

Comment (GwnDaan) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp:296]: Okay hmm I think I get it? Correct me if I'm wrong.

It now uses the `children` list of an object to store objects 'references' as if they were an actual child of the object. But then you don't actually use it as a child, but instead as a reference? I guess this would work yeah, though that also means we have to filter this `children` list when we want to serialize the object. Also a small downside is it requires at least 4 bytes extra storage as `ChildObjectData` also has the locations which are not used when acting as an objectId reference only. IMO not that big of a deal, but will it be worth the hassle?

To be completely honest, I think for _us in the future_, and also other new maintainers, it might be more consistent and clear if we just have for each attribute a dedicated member field of the class. Then no-one has to guess how it's stored and I think it's the most straight forward. 

I do like the way you are thinking of making the file smaller (and hence less code to maintain), but I think there might be too many edge cases attached to this, for it to worth it hmmm... What do you think?

File: isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp
Comment (ad3154) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp:304]: It's mostly there for `std::shared_ptr<VTObject> get_object_by_id(std::uint16_t objectID) const;` In my VT server I call this hundreds of times haha. My GUI objects actually inherit from each of these objects, and then also inherit from the GUI framework, so inside their `paint` routine I can get objects from the parent object pool easily.

If we could force people to interact with creating these objects only through some other class, then we could ensure the reference is valid... like, we could make the constructors badged, or private with a `friend` class of the object pool or something. IDK 🤔

File: isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp
Comment (ad3154) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp:296]: To be honest I hadn't really given a lot of thought to serializing the pool yet, so that is a good thing to point out. And yes, there is definitely a 4 byte RAM cost per reference.

I'm not particularly attached to my weird child object thing here, so I'm inclined to agree that maybe it'll be less weird/confusing if we just use dedicated fields for non-child references. I can make those changes

Comment (vbenso) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp:296]: Gentleman, 

I'm following the progress of this project, awesome work by the way.
About the VTOP, here goes my take. In the test's I've done here (this is a kind of weekend project for experimentation about the subject), the results using a serialized structure to store all the object in RAM, seem to be working. 
It's not as fast I wished it to be, but perhaps can be of some use for you. 
Also some feedback would be appreciated.
I'm using the following to parse a VTOP and, also, as a base to another class that can draw the elements based on the attributes of each into an LCD screen (actually at this development stage using an ESP32.

[opparser.zip](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/files/13429535/opparser.zip)

PS.: Sorry for not using such a modern C++ implementation, old habits die hard. I'll get there some day.

File: isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp
Comment (GwnDaan) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp:304]: Ohhh I see! I looked a bit in the VT server, and noticed in almost all places where you call `get_object_by_id` you also have a `VTServerManagerWorkingSet` to your disposal. Maybe it would be an idea to move this function to there? Since that object should also have access to the object pool. 

Also theoretically, one might want to handle objects independent from the complete pool. I think that is a little difficult right now by requiring an object pool for an object. 

> If we could force people to interact with creating these objects only through some other class, then we could ensure the reference is valid... like, we could make the constructors badged, or private with a friend class of the object pool or something.

I think instead of limiting the interaction with the objects, the approach above, where specific operations can only happen with some kind of middle-man that holds the reference as value/smart_pointer instead, might be a good tradeoff?

Comment (GwnDaan) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp:304]: > If we could force people to interact with creating these objects only through some other class, then we could ensure the reference is valid

I mean, that would only hold if the other class holds the pool as a value **and** lives longer than the vt-object. The last one you can't really guarantee without an extra dependency I think

File: isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp
Comment (ad3154) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp:296]: Hey there!

Not to get too off-topic on this PR, but here, take a look at [our parser](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/blob/2cbc71eee5beedb184be0ac4188fba515a4e14f0/isobus/src/isobus_virtual_terminal_server_managed_working_set.cpp#L145) as well for how we've been approaching it. It's maybe a bit heavy-weight for an ESP32 admittedly.

Some feedback I guess would be that it's mostly actually C! So, best-practices aside (such as having `using namespace` in a header file), I worry about how you're essentially just casting the pool data to the different types which would result in very strange things happening if the pool schema is not valid, possibly a hardfault/segfault. There might also be concerns between machines with different memory endianness? Though it's clearly working for ESP32 at least. Casting is efficient, when the pool is properly constructed yes, but intolerant of many OP issues which a server is obligated to report to a client I think unless I am missing something. `appendObject` seems like it could be the slowest part of parsing, lots of memcpy, malloc and free.

I think we are fairly close to making our VT GUI for PC available to the public, so be sure to keep a close eye on things! It has a few issues yet to work out, but I'd say it's maybe 80% complete. I've done some testing with 2 simultaneous working sets each with 10k+ objects and it seemed to hold up fairly well, I think.

File: isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp
Comment (ad3154) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp]: Removed stored references to both the object tree and colour table.

File: isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp
Comment (ad3154) [reply isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp]: Removed the function(s) mentioned, and refactored objects so that attributes of concern have dedicated setters and getters

**Review by GwnDaan — APPROVED** [review]

**Pull request merged**
